![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesor: Pablo Badilla, Diego Cortez
- Auxiliar: Melanie Peña Torres, Valentina Rojas Osorio
- Ayudante: Javiera Arévalo, Tamara Carrasco, Ignacio Reyes

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Álvaro Sifuentes Tasayco
- Estudiante n°2: Sebastián Morales Castillo

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

## Reglas

El proyecto debe realizarse al final del curso y su objetivo es poner a prueba el conocimiento de todas las unidades del curso. Este consta de 3 partes: **Informe de Análisis** (30%), **Construcción del Modelo** (40%) y **Despliegue del Modelo** (30%). Este notebook corresponde a la **Parte 1 - Informe de Análisis**.

### **Requisitos del proyecto**
- **Grupos**: Formar equipos de **2 personas**. 
- **Plazo de entrega**: Hasta el Viernes 10 de Julio
- **Consultas**: Cualquier duda fuera del horario de clases debe ser planteada en el foro correspondiente. Los mensajes enviados al equipo docente serán respondidos únicamente por este medio. Por favor, revisen las respuestas anteriores en el foro antes de realizar nuevas consultas.
- **Plagio**: La copia o reutilización no autorizada de trabajos de otros grupos está **estrictamente prohibida**. El incumplimiento de esta norma implicará la anulación inmediata del proyecto y una posible sanción académica.
- **Material permitido**: Pueden usar cualquier material del curso, ya sea notas, lecturas, códigos, o referencias proporcionadas por los docentes, que consideren útil para el desarrollo del proyecto.

---

## 📊 Distribución de Puntaje

Esta parte se evalúa en **escala de 1.0 a 7.0**, donde la nota se calcula como:

$$\text{Nota} = 1.0 + \text{puntaje obtenido}$$

El puntaje total repartible es de **6.0 puntos**, distribuidos así:

| Sección | Detalle | Puntaje |
| ------- | ------- | ------: |
| **Sección 1 - Entendimiento del Problema** | Preguntas conceptuales (P1–P7) | 2.0 |
| **Sección 2 · Subsección 1 - Datos Tabulares** | Código (1.0) + Preguntas (1.6) | 2.6 |
| **Sección 2 · Subsección 2 - Embeddings** | Código (0.6) + Preguntas (0.8) | 1.4 |
| **Total** | | **6.0** |

Cada pregunta y tarea indica su puntaje entre paréntesis, p. ej. _(0.3 pts)_.


Esta nota corresponde a la nota de la **Parte 1** que equivale a un **30%** de la nota del proyecto final. Para obtener la nota final, las notas individuales de cada parte se ponderan de la siguiente manera.

| Parte              | Ponderación |
|--------------------|------------:|
| **1 — Análisis**   | **30%**     |
| 2 — Modelo         | 40%         |
| 3 — Despliegue     | 30%         |

---

## 📖 Introducción 

<p align="center">
    <img src="./chaucherapp.jpg" alt="Alt Text" width="700">
</p>

La app chilena **ChaucherApp** ha emergido recientemente como una de las nuevas apps *FinTech* que apuesta por dominar en un competitivo mercado. Esta app ofrece una billetera digital que permite hacer transferencias, tarjetas digitales que permiten comprar, y servicios financieros como préstamos, seguros e inversiones.

Sin duda es sorprendente que **ChaucherApp** sea uno de los líderes de la industria fintech, siendo una empresa tan joven. En las palabras del CEO de la empresa, el Sr. Querty _'El gran éxito de ChaucherApp se debe a que entendimos la importancia de escuchar y responder eficientemente al cliente'_. En efecto, **ChaucherApp** fue diseñada desde el comienzo con una robusta arquitectura digital relacionada al **Soporte al cliente**.

Durante años, **ChaucherApp** ha logrado dar respuesta satisfactoria a cientos de tickets de soporte al cliente al mes, lo que le ha permitido inspirar confianza y prestigio. Por supuesto, esto también ha generado una gran cantidad de datos que pueden ser útiles para las siguientes etapas de crecimiento de la empresa. Y es que, a medida que el mercado se hace más competitivo, se hace más necesario aumentar la eficiencia de los procesos de respuesta de tickets, por lo que el Sr. Querty ha optado por **automatizar el proceso** incorporando inteligencia artificial.

Luego de intensa planificación ejecutiva y jornadas de analítica, han decidido que el primer paso para automatizar será el de desarrollar un modelo de **priorización de tickets** de soporte al cliente utilizando **modelos de lenguaje** y **aprendizaje de máquinas**. De esta manera podrán optimizar recursos humanos mientras se desarrollan otras herramientas.

Para lograr esta tarea, te han encargado analizar datos históricos de tickets de soporte al cliente que ya han sido manualmente priorizados, presentes en el archivo `data/tickets.parquet`. En él, encontrarás tickets con el contenido del ticket, categoría, canal, fecha y priorización, que puede ser `Baja`, `Media`, `Alta` o `Crítica`. Tu objetivo es utilizar esa data y los medios, algoritmos y estrategias que estimes convenientes para construir un **modelo clasificador** que sea capaz de predecir esta categoría en base al texto y atributos del ticket.

Con esta información, tus tareas a realizar son primero **analizar y entender los datos**, luego **construir un modelo de inferencia**, y finalmente **implementar ese modelo** para que el equipo de ejecutivos de ChaucherApp pueda utilizarlo.

---

# Sección 1 - Entendimiento del Problema

_(Total: 2.0 pts)_

El objetivo de esta sección es entender cuál será el problema que tendrás que resolver, incluyendo tanto la definición misma del problema (i.e., si es clasificación binaria, clasificación multiclase, regresión, clustering, otros...) como también las métricas, tanto de negocio como de evaluación, que usarás para medir el éxito de tu solución.

Para esto, primero responde las siguientes preguntas:


1. ¿Qué tipo de problema vamos a resolver? ¿Clasificación? ¿Regresión? ¿Clustering? ¿Reducción de dimensionalidad? ¿Ranking? ¿Recomendación? ¿Por qué? (utiliza la definición de cada uno de estos tipos de problemas más el enunciado para fundamentar). _(0.3 pts)_
2. ¿Por qué el negocio te solicitaría resolver este problema? ¿Por qué es importante? _(0.2 pts)_
3. ¿Por qué el negocio solicita generar un modelo predictivo para resolver este problema? ¿Por qué no lo pueden hacer, por ejemplo, por medio de código más palabras claves? _(0.2 pts)_
4. ¿Cuáles serían las métricas de evaluación relevantes para el negocio que buscarás optimizar? Aquí tienes la libertad de fundamentar lo que se te ocurra en base a tu entendimiento del negocio. Nombra al menos 3 y justifica la elección. _(0.3 pts)_
5. ¿Cuáles serían las métricas de evaluación de ML importantes para este problema? Para responder esto, primero da la interpretación de qué significa cada métrica (R^2, RMSE, MSE, etc.) / (Accuracy, Precision, Recall, F1) respecto a un problema binario Correcto / Incorrecto. Importante: Debes responder con la interpretación de qué significa cada una de estas métricas respecto al problema binario. _(0.5 pts)_
6. Luego, a partir de las interpretaciones anteriores de métricas más el conocimiento del negocio, elige cuál métrica es la más apropiada para cada clase en base a las prioridades del negocio. _(0.3 pts)_
7. Por último, ¿qué es la agregación macro? ¿Y la weighted? Y, ¿qué tipo de agregación utilizarías para promediar estas métricas de evaluación? _(0.2 pts)_

**Responde aquí**

> ...

---

# Sección 2 - Análisis de Datos

## Subsección 1 - Datos Tabulares

_(Total: 2.6 pts — código 1.0 + preguntas 1.6)_


La idea de esta segunda sección es que leas las fuentes de datos y las explores.


Partiremos primero con `tickets.parquet`. Utiliza la función de I/O asociada a este tipo de archivo para cargar los datos.

In [16]:
# Tu código aquí...
import pandas as pd
import plotly.express as px
from plotly.subplots import make_subplots

df_tickets = pd.read_parquet("data/tickets.parquet")

Ahora, analiza los datos de entrada usando las herramientas solicitadas _(código: 1.0 pts en total)_:


1. Utiliza `.info()` y `.describe()` para obtener una visualización rápida del dataframe. _(0.1 pts)_
2. Utiliza `histogram` de plotly express para explorar la distribución de la variable objetivo, y `value_counts` más operaciones entre series para calcular la cantidad y porcentaje de sus categorías. _(0.2 pts)_
3. Utiliza `histogram` más [subplots](https://plotly.com/python/subplots/) para observar la distribución de características numéricas y categóricas. Hint: puedes acceder con `fig.data[0]` a la traza de un histograma generado por `histogram`. _(0.2 pts)_
4. Utiliza `corr` + `imshow` para analizar la correlación entre variables numéricas. _(0.1 pts)_
5. Genera bins con `qcut` para `N_Caracteres_Ticket`, `Tiempo_de_Resolución_Horas` y `Usuario-Antiguedad_Cuenta_Dias`, con los valores `Bajo`, `Medio`, `Alto`. Luego, utiliza `crosstab` (normalizando usando `normalize` por filas) + `imshow` para analizar las contingencias entre `Categoría_Problema` y `Nivel_Prioridad`, `Canal_Ticket` y `Nivel_Prioridad`, `Usuario-Tipo_de_Cuenta` y `Nivel_Prioridad`, y los bins de `N_Caracteres_Ticket` y `Nivel_Prioridad`, `Tiempo_de_Resolución_Horas` y `Nivel_Prioridad`, y `Usuario-Antiguedad_Cuenta_Dias` y `Nivel_Prioridad`. Grafica en subplots. _(0.2 pts)_
6. Por último, agrega la cantidad de reclamos por mes y `Nivel_Prioridad` (crea una nueva columna llamada `Fecha_Envío_Mes` usando el módulo de datetimes de pandas) para obtener un conteo de cantidad de reclamos según mes-nivel de prioridad, y utiliza `line` para generar un gráfico de línea en donde cada `Nivel_Prioridad` tenga su propia serie. _(0.2 pts)_

In [17]:
# Tu código aquí...
# Recuerda que puedes usar múltiples celdas.
# Pregunta 1: Información general del dataframe y Estadísticos descriptivos de las variables numéricas
df_tickets.info()
df_tickets.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   Id_Ticket                       2000 non-null   object
 1   Asunto_Ticket                   2000 non-null   object
 2   Contenido_Ticket                2000 non-null   object
 3   N_Caracteres_Ticket             2000 non-null   int64 
 4   Canal_Ticket                    2000 non-null   object
 5   Categoría_Problema              2000 non-null   object
 6   Usuario-Tipo_de_Cuenta          2000 non-null   object
 7   Usuario-Antiguedad_Cuenta_Dias  2000 non-null   int64 
 8   Fecha_Envío                     2000 non-null   object
 9   Tiempo_de_Resolución_Horas      2000 non-null   int64 
 10  Nivel_Prioridad                 2000 non-null   object
dtypes: int64(3), object(8)
memory usage: 172.0+ KB


,N_Caracteres_Ticket,Usuario-Antiguedad_Cuenta_Dias,Tiempo_de_Resolución_Horas
count,2000.000000,2000.000000,2000.000000
mean,717.699000,121.743000,18.718000
std,209.355096,142.339366,18.841788
min,251.000000,4.000000,1.000000
25%,564.000000,42.000000,5.000000
50%,697.000000,78.000000,13.000000
75%,850.250000,143.000000,26.000000
max,1831.000000,1200.000000,183.000000


In [18]:
# Pregunta 2: Distribución de la variable objetivo
target_col = "Nivel_Prioridad"

conteo = df_tickets[target_col].value_counts()
porcentaje = conteo / conteo.sum() * 100

resumen_objetivo = pd.DataFrame({"Cantidad": conteo, "Porcentaje": porcentaje})

print(resumen_objetivo)

fig = px.histogram(df_tickets, x=target_col)

fig.show()

                 Cantidad  Porcentaje
Nivel_Prioridad                      
Media                 789       39.45
Baja                  723       36.15
Alta                  350       17.50
Critica               138        6.90


In [19]:
# Pregunta 3: Distribución de características numéricas y categóricas

num_cols = ["N_Caracteres_Ticket", "Tiempo_de_Resolución_Horas", "Usuario-Antiguedad_Cuenta_Dias"]

cat_cols = ["Categoría_Problema", "Canal_Ticket", "Usuario-Tipo_de_Cuenta"]

fig = make_subplots(rows=2, cols=3, subplot_titles=num_cols + cat_cols)

for i, col in enumerate(num_cols):
    hist = px.histogram(df_tickets, x=col)
    fig.add_trace(hist.data[0], row=1, col=i + 1)

for i, col in enumerate(cat_cols):
    hist = px.histogram(df_tickets, x=col)
    fig.add_trace(hist.data[0], row=2, col=i + 1)

fig.show()

In [20]:
# Pregunta 4: Correlación entre variables numéricas

corr = df_tickets[num_cols].corr()

fig = px.imshow(corr, text_auto=True)

fig.show()

In [21]:
# Pregunta 5: Bins, contingencias y subplots

df_tickets["Bin_N_Caracteres_Ticket"] = pd.qcut(
    df_tickets["N_Caracteres_Ticket"], q=3, labels=["Bajo", "Medio", "Alto"]
)

df_tickets["Bin_Tiempo_de_Resolución_Horas"] = pd.qcut(
    df_tickets["Tiempo_de_Resolución_Horas"], q=3, labels=["Bajo", "Medio", "Alto"]
)

df_tickets["Bin_Usuario_Antiguedad_Cuenta_Dias"] = pd.qcut(
    df_tickets["Usuario-Antiguedad_Cuenta_Dias"], q=3, labels=["Bajo", "Medio", "Alto"]
)

variables = [
    "Categoría_Problema",
    "Canal_Ticket",
    "Usuario-Tipo_de_Cuenta",
    "Bin_N_Caracteres_Ticket",
    "Bin_Tiempo_de_Resolución_Horas",
    "Bin_Usuario_Antiguedad_Cuenta_Dias",
]

fig = make_subplots(rows=2, cols=3, subplot_titles=variables)

for i, col in enumerate(variables):
    tabla = pd.crosstab(df_tickets[col], df_tickets["Nivel_Prioridad"], normalize="index")

    heatmap = px.imshow(tabla, text_auto=True)

    fig.add_trace(heatmap.data[0], row=i // 3 + 1, col=i % 3 + 1)

fig.show()

In [22]:
# Pregunta 6: Reclamos por mes y nivel de prioridad

df_tickets["Fecha_Envío"] = pd.to_datetime(df_tickets["Fecha_Envío"])

df_tickets["Fecha_Envío_Mes"] = df_tickets["Fecha_Envío"].dt.to_period("M").astype(str)

reclamos_mes = df_tickets.groupby(["Fecha_Envío_Mes", "Nivel_Prioridad"]).size().reset_index(name="Cantidad")

fig = px.line(reclamos_mes, x="Fecha_Envío_Mes", y="Cantidad", color="Nivel_Prioridad")

fig.show()

Ahora, responde las siguientes preguntas _(preguntas: 1.6 pts en total)_:

1. ¿Cuántos datos son? ¿Hay nulos? _(0.1 pts)_
2. ¿Qué columnas tenemos disponibles? ¿Qué tipo de dato es cada columna? - Tanto en tipo de dato (str, int, float, etc.) como también en su contenido (texto libre, categorías, entero, número decimal, fechas, índices) y su relación. Responde en una tabla con: Columna, Tipo de dato, Tipo de contenido y Descripción. _(0.2 pts)_
3. ¿Cómo se distribuyen las etiquetas? ¿Hay desbalance? Incluye porcentajes en la respuesta. _(0.2 pts)_
4. ¿Cómo afecta la distribución de clases a las respuestas dadas en la sección 1? ¿Sería necesario editar o corregir algo de estas? Si es así, edita las respuestas de la sección 1 y menciona qué edición hiciste aquí. _(0.1 pts)_
5. ¿Cómo se distribuyen las variables numéricas? ¿Con cuáles distribuciones las podríamos relacionar? ¿Cuál es la interpretación de estas? ¿Existen outliers? Explica por cada variable numérica. _(0.2 pts)_
6. ¿Existe alguna correlación entre las variables numéricas? _(0.1 pts)_
7. ¿Existe alguna relación interesante entre las categorías generadas y las categorías de la variable objetivo? _(0.2 pts)_
8. ¿Existe alguna relación entre el tiempo y la variable objetivo? ¿Sería recomendable utilizar esta variable? _(0.1 pts)_
9. Pensemos ahora por un momento en el modelo en modo inferencia/producción. ¿Existe alguna variable que no vaya a estar disponible en ese momento? Si es así, ¿cómo se llama este fenómeno? ¿Si no hubiésemos meditado esto, qué hubiese sucedido? _(0.2 pts)_
10. Para generar los bins de las variables numéricas se utilizó `qcut` (cuantiles) en lugar de `cut` (ancho fijo). ¿Qué habría pasado con el heatmap de contingencia de `Tiempo_de_Resolución_Horas` si hubieras usado `cut`? ¿Por qué `qcut` es preferible cuando la variable está muy sesgada? _(0.2 pts)_

**Responde aquí**

Respuesta 1: El dataframe contiene 2000 datos, es decir, 2000 filas y 11 columnas. A partir de `.info()` se observa que todas las columnas tienen 2000 valores no nulos, por lo que no hay valores nulos en el conjunto de datos.
Además, el dataframe posee 3 columnas numéricas de tipo `int64` y 8 columnas de tipo `object`.

# Respuesta 2

| Columna | Tipo de dato | Tipo de contenido | Descripción |
|---|---|---|---|
| Id_Ticket | object | Identificador | Identificador único de cada ticket. Sirve como referencia del caso, pero no entrega información predictiva directa. |
| Asunto_Ticket | object | Texto libre | Texto corto asociado al asunto o título del reclamo, este puede contener información semántica útil sobre el problema reportado. |
| Contenido_Ticket | object | Texto libre | Descripción principal del ticket escrita por el usuario, contiene el detalle del problema reportado. |
| N_Caracteres_Ticket | int64 | Entero | Cantidad de caracteres del ticket enviado por el usuario. |
| Canal_Ticket | object | Categoría | Canal por el cual ingresó el ticket, por ejemplo Whatsapp, Correo o Página Web. |
| Categoría_Problema | object | Categoría | Tipo de problema asociado al ticket, por ejemplo Cuenta, Otro, Fraude, Técnica, Pregunta general o Cobros. |
| Usuario-Tipo_de_Cuenta | object | Categoría | Tipo de cuenta del usuario, por ejemplo Free, Premium o Business. |
| Usuario-Antiguedad_Cuenta_Dias | int64 | Entero | Antigüedad de la cuenta del usuario medida en días. |
| Fecha_Envío | object | Fecha | Fecha en que fue enviado el ticket. Aunque aparece como object, su contenido corresponde a una fecha y puede transformarse a datetime. |
| Tiempo_de_Resolución_Horas | int64 | Entero | Cantidad de horas que demoró la resolución del ticket, esta es una variable temporal posterior a la atención del caso. |
| Nivel_Prioridad | object | Categoría objetivo | Variable objetivo del problema. Indica la prioridad del ticket: Baja, Media, Alta o Critica. |

# Respuesta 3:

La variable objetivo Nivel_Prioridad se distribuye de la siguiente forma: la clase Media tiene 789 casos, equivalentes al 39.45% del total; la clase Baja tiene 723 casos, equivalentes al 36.15%; la clase Alta tiene 350 casos, equivalentes al 17.50%; y la clase Critica tiene 138 casos, equivalentes al 6.90%.
Si existe desbalance de clases, ya que las clases Media y Baja concentran la mayor parte de los datos, sumando aproximadamente 75.6% del total. Mientras que las clases Alta y especialmente Critica tienen una representación mucho menor.
Esto es relevante porque un modelo podría tender a predecir principalmente las clases mayoritarias y aun así obtener un rendimiento global aparentemente bueno. Sin embargo, podría fallar en las clases más importantes desde el punto de vista operacional, como Alta y Critica.

# Respuesta 4

La distribución de clases afecta directamente la forma en que se debe plantear y evaluar el problema. Como existe desbalance, no sería suficiente utilizar únicamente métricas globales como accuracy, ya que un modelo podría obtener buenos resultados prediciendo principalmente las clases mayoritarias Media y Baja, pero tener un mal desempeño en las clases minoritarias Alta y Critica.
Por esto, sí sería necesario corregir o complementar las respuestas de la sección 1 si ahí se asumió que todas las clases estaban equilibradas o si se consideró accuracy como métrica principal. La edición realizada consiste en agregar que, debido al desbalance de clases, se deben priorizar métricas que evalúen el rendimiento por clase, como macro F1-score, balanced accuracy, recall por clase y matriz de confusión.
También se debe considerar el uso de particiones estratificadas para mantener proporciones similares de clases en entrenamiento, validación y test.

# Respuesta 5

Las variables numéricas disponibles son N_Caracteres_Ticket, Usuario-Antiguedad_Cuenta_Dias y Tiempo_de_Resolución_Horas. La variable N_Caracteres_Ticket presenta una distribución con un único punto máximo, concentrada principalmente entre 500 y 900 caracteres. Su media es aproximadamente de 717.7 caracteres y su mediana es 697, por lo que la mayoría de los tickets tiene una extensión intermedia. Esta distribución podría relacionarse con una distribución aproximadamente normal, aunque presenta una cola hacia la derecha. Sí existen posibles outliers, especialmente tickets muy largos, ya que el valor máximo alcanza 1831 caracteres.
La variable Tiempo_de_Resolución_Horas presenta una distribución fuertemente sesgada hacia la derecha. La mayoría de los tickets se resuelve en pocas horas, pero existen algunos casos que tardan mucho más. Su media es de 18.7 horas y su mediana es de 13 horas, lo que confirma el sesgo positivo. Esta variable podría relacionarse con una distribución exponencial o log-normal, debido a que concentra muchos valores pequeños y pocos valores muy altos. Existen outliers claros, con tiempos de resolución que llegan hasta 183 horas.
La variable Usuario-Antiguedad_Cuenta_Dias también está muy sesgada hacia la derecha. La mayoría de los usuarios tiene poca antigüedad, pero existen algunos usuarios con cuentas mucho más antiguas. Su media es de 121.7 días y su mediana es de 78 días, lo que muestra que los valores altos empujan la media hacia arriba. Esta variable también podría relacionarse con una distribución exponencial o log-normal. Existen outliers importantes, ya que el máximo alcanza 1200 días.

# Respuesta 6

No se observa una correlación fuerte entre las variables numéricas, pero la mayor correlación aparece entre N_Caracteres_Ticket y Tiempo_de_Resolución_Horas, con un valor aproximado de 0.194, la cual es una correlación positiva débil, por lo que podría sugerir que tickets más largos tienden a tardar un poco más en resolverse, pero la relación no es suficientemente fuerte como para considerarla determinante.
Las demás correlaciones son cercanas a cero, donde la correlación entre N_Caracteres_Ticket y Usuario-Antiguedad_Cuenta_Dias es aproximadamente 0.065, mientras que la correlación entre Tiempo_de_Resolución_Horas y Usuario-Antiguedad_Cuenta_Dias es aproximadamente 0.025.
Por lo tanto, no parece existir una relación lineal relevante entre las variables numéricas.

# Respuesta 7

Sí, se observan relaciones interesantes entre algunas variables categóricas y Nivel_Prioridad.
En Categoría_Problema, la categoría Fraude destaca por tener una proporción alta de tickets Alta y Critica, lo cual tiene sentido, ya que los problemas asociados a fraude suelen requerir atención más urgente. En cambio, Pregunta general se concentra principalmente en prioridad Baja, sin presencia relevante de tickets Alta o Critica.
En Canal_Ticket, los tickets que ingresan por Correo muestran una mayor proporción de prioridades Alta y Critica en comparación con otros canales. Por otro lado, Whatsapp concentra más casos de prioridad Baja y Media.
En Usuario-Tipo_de_Cuenta, los usuarios Business presentan una proporción considerablemente mayor de tickets Alta y Critica, mientras que los usuarios Free concentran más tickets de prioridad Baja y Media. Esto sugiere que el tipo de cuenta puede estar relacionado con la prioridad asignada.
Respecto a los bins numéricos, Bin_N_Caracteres_Ticket muestra una relación clara: los tickets más largos tienen mayor proporción de prioridad Alta y Critica, mientras que los tickets más cortos se concentran principalmente en prioridad Baja. Esto indica que la extensión del ticket podría aportar información predictiva.
Para Bin_Tiempo_de_Resolución_Horas, los tickets con mayor tiempo de resolución tienden a presentar una mayor proporción de prioridad Alta y Critica. Sin embargo, esta variable debe analizarse con cuidado, porque probablemente no estará disponible al momento de hacer una predicción en producción.
Finalmente, en Bin_Usuario_Antiguedad_Cuenta_Dias, los usuarios con menor antigüedad presentan una mayor proporción de tickets Alta y Critica. En cambio, los usuarios con antigüedad media tienden a concentrarse más en prioridad Baja.

# Respuesta 8

A partir del gráfico de reclamos por mes y Nivel_Prioridad, se observa que las clases mantienen un patrón relativamente estable a lo largo del tiempo. Las categorías Media y Baja suelen ser las más frecuentes durante casi todo el período, mientras que Critica se mantiene como la menos frecuente. Existen fluctuaciones mensuales, pero no se aprecia una tendencia temporal fuerte ni un cambio claro en la distribución de prioridades.
La variable Fecha_Envío sí estaría disponible al momento de recibir un ticket, por lo que podría utilizarse en producción si se transforma adecuadamente. Por ejemplo, se podrían extraer variables como mes, día de la semana o alguna señal estacional. Sin embargo, según el gráfico observado, no parece ser una variable con una señal muy fuerte.
Por lo tanto, no sería recomendable usar la fecha directamente como texto o identificador temporal crudo. Podría utilizarse con cuidado mediante características derivadas, pero habría que validar si realmente aporta al rendimiento del modelo y si no genera sobreajuste a meses específicos del conjunto de entrenamiento.

# Respuesta 9 

Sí. La variable Tiempo_de_Resolución_Horas probablemente no estará disponible al momento de hacer la predicción en producción, porque corresponde a información que se conoce después de que el ticket fue resuelto.
Este fenómeno se llama data leakage o fuga de información. En este caso corresponde a una fuga temporal, porque se estaría usando información futura para predecir la prioridad de un ticket en el momento de su ingreso.
Si no se hubiese considerado esto, el modelo podría haber aprendido una relación artificialmente buena usando información que no existe al momento real de predicción. Como consecuencia, el rendimiento en validación o test podría verse inflado, pero el modelo no funcionaría correctamente en producción o directamente no podría utilizarse con esa variable.

# Respuesta 10

Si se hubiese usado cut, los bins se habrían creado con intervalos de igual ancho. Como Tiempo_de_Resolución_Horas está muy sesgada hacia la derecha, la mayor parte de los datos habría quedado concentrada en el primer intervalo, mientras que los intervalos superiores habrían tenido muy pocas observaciones.
Esto habría producido un heatmap de contingencia menos informativo y posiblemente inestable, porque algunas filas tendrían muy pocos casos. En ese escenario, los porcentajes por fila podrían estar dominados por pocas observaciones y no representar bien el comportamiento general de la variable.
qcut es preferible cuando la variable está muy sesgada porque divide los datos según cuantiles, generando grupos con cantidades similares de observaciones. Esto permite comparar mejor las proporciones de Nivel_Prioridad entre los grupos Bajo, Medio y Alto, evitando bins casi vacíos o demasiado desbalanceados.

---

## Subsección 2 - Embeddings

_(Total: 1.4 pts — código 0.6 + preguntas 0.8)_

En esta sección trabajarás con embeddings sobre los Asuntos + Descripciones de tickets precalculados con el modelo `gemini-embeddings-001`. Ahora, nuevamente utilizando las herramientas de lectura de parquet _(código: 0.6 pts en total)_:

1. Lee el archivo que contiene los embeddings (`data/embeddings.parquet`). _(0.1 pts)_
2. Reduce la dimensionalidad de los embeddings con umap-learn a dos dimensiones y agrega a `df_tickets` las nuevas dimensiones. Asegura tu proyección utilizando semillas, así luego, al momento de la revisión, se podrán replicar los resultados. _(0.3 pts)_
3. Grafica un scatter plot con las proyecciones y pinta por `Nivel_Prioridad` (utiliza el código dado).
4. Grafica un nuevo scatter plot con las proyecciones y pinta por `Categoría_Problema`. _(0.2 pts)_

In [23]:
# Tu código aquí...
# Recuerda que puedes usar múltiples celdas.



In [24]:
import textwrap


def wrap_text(text, width=40):
    return "<br>".join(textwrap.wrap(text, width=width))


df_tickets["Contenido_Ticket_Wrapped"] = df_tickets["Contenido_Ticket"].apply(lambda x: wrap_text(x, 80))

fig_proj = px.scatter(
    df_tickets,
    x="x",
    y="y",
    color="Nivel_Prioridad",
    title="Proyecciones de los embeddings respecto a clase",
    hover_data=[
        "Canal_Ticket",
        "N_Caracteres_Ticket",
        "Categoría_Problema",
        "Usuario-Tipo_de_Cuenta",
        "Usuario-Antiguedad_Cuenta_Dias",
        "Contenido_Ticket_Wrapped",
    ],
)
fig_proj.show()

ValueError: Value of 'x' is not the name of a column in 'data_frame'. Expected one of ['Id_Ticket', 'Asunto_Ticket', 'Contenido_Ticket', 'N_Caracteres_Ticket', 'Canal_Ticket', 'Categoría_Problema', 'Usuario-Tipo_de_Cuenta', 'Usuario-Antiguedad_Cuenta_Dias', 'Fecha_Envío', 'Tiempo_de_Resolución_Horas', 'Nivel_Prioridad', 'Bin_N_Caracteres_Ticket', 'Bin_Tiempo_de_Resolución_Horas', 'Bin_Usuario_Antiguedad_Cuenta_Dias', 'Fecha_Envío_Mes', 'Contenido_Ticket_Wrapped'] but received: x

Ahora, responde las siguientes preguntas _(preguntas: 0.8 pts en total)_:

1. ¿Qué es un embedding? ¿Qué representan? A grandes rasgos, ¿cómo se calculan? Hint: consulta sobre `gemini-embedding-001`. _(0.2 pts)_
2. ¿Por qué los embeddings tienen el potencial de ser mejores para un modelo predictivo que analizar las palabras individuales? _(0.1 pts)_
3. ¿Cuál podría ser una desventaja de utilizar embeddings en vez de utilizar directamente palabras? _(0.1 pts)_
4. ¿Existen otros tipos de embeddings? ¿Qué podrían representar? _(0.1 pts)_
5. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Nivel_Prioridad`? _(0.1 pts)_
6. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Categoría_Problema`? _(0.1 pts)_
7. Las proyecciones de UMAP (`x`, `y`) quedaron agregadas a `df_tickets`. ¿Sería buena idea usar estas 2 dimensiones como features de entrada del modelo predictivo, en lugar de los embeddings originales? Justifica pensando en la información que se pierde al reducir a 2D. _(0.1 pts)_

**Responde aquí**

> ...